charger/semp/types.go

```diff
@@ -0,0 +1,264 @@
+
+// DeviceStatus represents SEMP device status
+type DeviceStatus struct {
+	DeviceID          string    `xml:"DeviceId"`
+	Status            string    `xml:"Status"`
+	EMSignalsAccepted bool      `xml:"EMSignalsAccepted"`
+	PowerInfo         PowerInfo `xml:"PowerConsumption>PowerInfo"`
+}
+
+// PowerInfo represents SEMP power information
+type PowerInfo struct {
+	AveragePower      int `xml:"AveragePower"`
+	Timestamp         int `xml:"Timestamp"`
+	AveragingInterval int `xml:"AveragingInterval"`
+}
+
+// DeviceControl represents SEMP device control message
+type DeviceControl struct {
+	DeviceID                    string `xml:"DeviceId"`
+	On                          bool   `xml:"On"`
+	RecommendedPowerConsumption int    `xml:"RecommendedPowerConsumption"`
+	Timestamp                   int    `xml:"Timestamp"`
+}
+
+// PlanningRequest represents SEMP planning request
+type PlanningRequest struct {
+	Timeframe []Timeframe `xml:"Timeframe"`
+}
+
+// Timeframe represents SEMP timeframe
+type Timeframe struct {
+	DeviceID            string `xml:"DeviceId"`
+	EarliestStart       int    `xml:"EarliestStart"`
+	LatestEnd           int    `xml:"LatestEnd"`
+	MinRunningTime      *int   `xml:"MinRunningTime,omitempty"`
+	MaxRunningTime      *int   `xml:"MaxRunningTime,omitempty"`
+	MinEnergy           *int   `xml:"MinEnergy,omitempty"`
+	MaxEnergy           *int   `xml:"MaxEnergy,omitempty"`
+	MaxPowerConsumption *int   `xml:"MaxPowerConsumption,omitempty"`
+	MinPowerConsumption *int   `xml:"MinPowerConsumption,omitempty"`
+}
+
+// Device2EM represents the device to energy manager message
+type Device2EM struct {
+	XMLName         xml.Name          `xml:"Device2EM"`
+	Xmlns           string            `xml:"xmlns,attr"`
+	DeviceInfo      []DeviceInfo      `xml:"DeviceInfo,omitempty"`
+	DeviceStatus    []DeviceStatus    `xml:"DeviceStatus,omitempty"`
+	PlanningRequest []PlanningRequest `xml:"PlanningRequest,omitempty"`
+}
+
+// EM2Device represents the energy manager to device message
+type EM2Device struct {
+	XMLName       xml.Name        `xml:"EM2Device"`
+	Xmlns         string          `xml:"xmlns,attr"`
+	DeviceControl []DeviceControl `xml:"DeviceControl,omitempty"`
+}
+
+// MarshalXML marshals XML into an io.ReadSeeker
+func MarshalXML(data interface{}) io.ReadSeeker {
+	if data == nil {
+		return nil
+	}
+
+	body, err := xml.Marshal(data)
+	if err != nil {
+		return &errorReader{err: err}
+	}
+
+	return bytes.NewReader(body)
+}
+
+// errorReader wraps an error with an io.Reader
+type errorReader struct {
+	err error
+}
+
+func (r *errorReader) Read(p []byte) (int, error) {
+	return 0, r.err
+}
+
+func (r *errorReader) Seek(offset int64, whence int) (int64, error) {
+	return 0, r.err
+}
+
+// Client represents a SEMP HTTP client with helper methods
+type Client struct {
+	helper   *request.Helper
+	uri      string
+	deviceID string
+}
+
+// NewClient creates a new SEMP client
+func NewClient(helper *request.Helper, uri, deviceID string) *Client {
```

Comment:
Der Teil sollte in client.go oder connection.go

Reply:
Ja, das geht in der Tat besser.